raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2669,7 +2721,14 @@ private long pollResigned(long currentTimeMs) {
             // until either the shutdown expires or an election bumps the epoch
             stateTimeoutMs = shutdown.remainingTimeMs();
         } else if (state.hasElectionTimeoutExpired(currentTimeMs)) {
-            transitionToCandidate(currentTimeMs);
+            if (quorum.isVoter()) {
+                transitionToCandidate(currentTimeMs);
+            } else {
+                // It is posible that the old leader is not a voter in the new voter set.
```

Comment:
Should there be a log4j message here?

Reply:
KRaft already logs a message at info level for every state transition:
https://github.com/apache/kafka/blob/trunk/raft/src/main/java/org/apache/kafka/raft/QuorumState.java#L577